lib/fluent/plugin/in_tail.rb

```diff
@@ -537,86 +535,144 @@ def swap_state(pe)
         end
 
         def on_change(prev, cur)
           @callback.call
         rescue
           # TODO log?
-          @log.error $!.to_s
-          @log.error_backtrace
+          @watcher.log.error $!.to_s
+          @watcher.log.error_backtrace
+        end
+      end
+
+
+      class FIFO
+        def initialize(from_encoding, encoding)
+          @from_encoding = from_encoding
+          @encoding = encoding
+          @buffer = ''.force_encoding(from_encoding)
+          @eol = "\n".encode(from_encoding).freeze
+        end
+
+        attr_reader :from_encoding, :encoding, :buffer
+
+        def <<(chunk)
+          orig_encoding = chunk.encoding
+          chunk.force_encoding(from_encoding)
+          @buffer << chunk
+          chunk.force_encoding(orig_encoding)
+        end
+
+        def convert(s)
+          if @from_encoding == @encoding
+            s
+          else
+            c = Encoding::Converter.new(@from_encoding, @encoding)
+            c.convert(s) + c.finish()
+          end
+        end
+
+        def next_line
+          idx = @buffer.index(@eol)
+          convert(@buffer.slice!(0, idx + 1)) unless idx.nil?
+        end
+
+        def size
+          @buffer.bytesize
         end
       end
 
       class IOHandler
-        def initialize(io, pe, log, read_lines_limit, first = true, &receive_lines)
-          @log = log
-          @log.info "following tail of #{io.path}" if first
-          @io = io
-          @pe = pe
-          @read_lines_limit = read_lines_limit
+        def initialize(watcher, &receive_lines)
+          @watcher = watcher
           @receive_lines = receive_lines
-          @buffer = ''.force_encoding('ASCII-8BIT')
+          @fifo = FIFO.new(@watcher.from_encoding || Encoding::ASCII_8BIT, @watcher.encoding || Encoding::ASCII_8BIT)
           @iobuf = ''.force_encoding('ASCII-8BIT')
           @lines = []
+          @io = nil
+          @watcher.log.info "following tail of #{@watcher.path}"
         end
 
-        attr_reader :io
-        attr_accessor :pe
-
         def on_notify
-          begin
-            read_more = false
-
-            if @lines.empty?
-              begin
-                while true
-                  if @buffer.empty?
-                    @io.readpartial(2048, @buffer)
-                  else
-                    @buffer << @io.readpartial(2048, @iobuf)
-                  end
-                  while idx = @buffer.index("\n".freeze)
-                    @lines << @buffer.slice!(0, idx + 1)
-                  end
-                  if @lines.size >= @read_lines_limit
-                    # not to use too much memory in case the file is very large
-                    read_more = true
-                    break
+          with_io do |io|
+            begin
+              read_more = false
+
+              if !io.nil? && @lines.empty?
+                begin
+                  while true
+                    @fifo << io.readpartial(2048, @iobuf)
```

Comment:
That's exactly what I was wondering about.  I just did so as the original code does.